Feature/render

redsteadz opened the pull request:
Implementation for issue: #2 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Introduced a window management class for easier window creation and control.
  * Added a sprite struct to encapsulate texture, source rectangle, tint, and layer information for rendering.
  * Implemented a renderer class supporting offscreen rendering, sprite draw queue, and layer-based sorting.
  * Added a rendering system function stub for future ECS integration.

* **Refactor**
  * Simplified the main application loop to focus on window and rendering setup, removing entity/component management.

* **Chores**
  * Updated configuration to improve C++ header recognition in the code editor.
  * Expanded core header to include new rendering-related modules.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

redsteadz committed:
feat: implement rendering system with fullscreen toggle and window management

redsteadz committed:
feat: add scaling mode and rendering updates for dynamic window resizing

redsteadz committed:
feat: implement Render and TextureManager classes with sprite rendering support

redsteadz committed:
fix: dereference sprite texture in DrawSprite method for correct rendering

DarkiCraft committed:
fix: directory structure + minor bug

DarkiCraft committed:
Merge branch 'feature/Render' of https://github.com/DarkiCraft/S.E.C.S.Y into feature/Render

DarkiCraft committed:
fix: wrong include

redsteadz committed:
feat: add sprite rendering support with RenderSystem and update sprite handling

redsteadz committed:
fix: Placeholder rect in case of assest not being found

darkicraft commented on src/SECSY/Render/Render.hpp:
```diff
@@ -0,0 +1,165 @@
+    CloseWindow();
+  }
+
+  void ToggleFullscreenMode() {
```
We gotta remove fullscreen, it's SUPER buggy on Windows. We can go boderless windowed (the top pane is cut of or disabled) or just a regular resizable window.

redsteadz replied:
Alrighty, we sure can. It's actually neat that you on Windows and me on Linux, technically testing out both systems haha

redsteadz committed:
fix: remove redundant render.BeginRendering() call and update window size handling in Render class

redsteadz commented on main.cpp:
```diff
@@ -39,6 +40,23 @@ int main() {
+    if (IsKeyPressed(KEY_F11)) {
+      render.ToggleFullscreenMode();
+    }
+  }
```
@DarkiCraft Doing this causes MAJOR buggy-ness in rendering and dynamic screen resizing

DarkiCraft committed:
refactor: `Render/` and `Platform/`

Abstracted window management winto `Platform/Window.hpp`. (Very simple implementation for now)

Made a dedicated `Render/Renderer.hpp` to handle layers,  draw queues and window scaling. LETTERBOX vs. PIXELPERFECT are removed for now.

`Sprite` now lives under `Render/Components.hpp`, alongside `Texture` in the future.

A very basic `RenderSystem` implementation is provided which will be uncommented when `refactor/ECS` is complete

DarkiCraft committed:
fix: added includes to `SECSY.hpp` to reflect previous commit

DarkiCraft commented on src/SECSY/Render/Renderer.hpp:
```diff
@@ -0,0 +1,97 @@
+        WHITE);
+
+    EndDrawing();
+  }
```
1. If the draw queue is cleared, would that not lead to missing sprites/artifacting/flickering across frames? If so, what is a better solution than outright removing this?

2. Can you explain the conflict you mentioned here:
> It calls `BeginDrawing()`/`EndDrawing()` internally, which conflicts with external rendering control.

darkicraft replied:
@coderabbitai @coderabbit @rabbitai @rabbit

(I don't know the correct @ for this, it's been so long T_T)

Firstly, `main.cpp` doesn't call `BeginDrawing()`/`EndDrawing()`, it calls `Renderer.Begin()` and `Renderer.End()` and between them, a sprite draw command is issued via `Renderer.Submit()`.

I understand externalizing `BeginDrawing()` and `EndDrawing()` and I thought about moving it to `Platform/` (maybe in `Windows.hpp`).

That leaves our `SECSY::Renderer` class to only have a single `::Submit()` function.

I am confused on how this will all work because then it will be harder to draw at the correct times. If `BeginDrawing()` and `EndDrawing()` is external, then we have no indication within `SECSY::Renderer` on when to draw.

Suppose:

Begin() called -> End() called -> Draw command submitted -> crash??? (drawing after EndDrawing())

Please explain how to move this forward, or if the current implementation is okay.

DarkiCraft replied:
We cannot use this implementation because this nests `BeginDrawing()`/`EndDrawing()` (internal to `window.BeginFrame()`/`window.EndFrame()` respectively) with `BeginTextureMode()`/`EndTextureMode()` (internal to `renderer.Begin()`/`renderer.End()` respectively), which is undefined behaviour in `raylib`.

darkicraft replied:
@coderabbitai

We cannot use this implementation because this nests `BeginDrawing()`/`EndDrawing()` (internal to `window.BeginFrame()`/`window.EndFrame()` respectively) with `BeginTextureMode()`/`EndTextureMode()` (internal to `renderer.Begin()`/`renderer.End()` respectively), which is undefined behaviour in raylib.

darkicraft replied:
@coderabbitai
Can you please cite the source on the "correct nesting order"?

DarkiCraft committed:
fix: disallow copying for `SECSY::Renderer` and `SECSY::Window` plus minor fixes
